UBL - Hilbert space-filling curve probing sequence

marciot opened the pull request:
The current probe pattern used by UBL causes a lot of extra travel and does not make sense for a rectangular bed, as noted in https://github.com/MarlinFirmware/Marlin/issues/16950.

This PR solves the problem by adding an additional option called UBL_HILBERT_CURVE which probes the bed using a [Hilbert space-filling curve](https://en.wikipedia.org/wiki/Hilbert_curve). This pattern was chosen over a simple zig-zag because it:

- Keeps travel moves to a minimum
- Probes all points physically close to each other before moving to a further region of the bed
- Looks aesthetically pleasing on the FTDI Eve Touch UI (see attached MOV file for an 8x8 grid)
- Pattern can be cropped to any grid size while preserving above advantages
- Marketing types can now claim the world's first **fractal bed leveling technology** :laughing:

https://user-images.githubusercontent.com/6752228/111838974-b151ef00-88bf-11eb-995f-0a256179a393.mp4

marciot committed:
Implemented Hilbert curve probe point ordering

GhostlyCrowd commented:
Brilliant, this will probably save a lot of time on my 100 point UBL

GhostlyCrowd commented:
G26 mesh validation to follow UBL_HILBERT_CURVE to reduce movement would be excellent as well.

Roxy-3D commented:
> G26 mesh validation to follow UBL_HILBERT_CURVE to reduce movement would be excellent as well.

Yes!  What ever is done for UBL's G29 should be done for G26 also.

GhostlyCrowd commented:
This is working for me, G26 following the same curve would be glorious.

X-Ryl669 commented on Marlin/src/feature/bedlevel/ubl/ubl_G29.cpp:
```diff
@@ -1298,6 +1302,55 @@ mesh_index_pair unified_bed_leveling::find_closest_mesh_point_of_type(const Mesh
 
+#if ENABLED(UBL_HILBERT_CURVE)
+   void unified_bed_leveling::hilbert(mesh_index_pair &pt, float x, float y, float xi, float xj, float yi, float yj, int n) {
+      /* Hilbert space filling curve implementation
```
Why do you use float here ?

marciot replied:
I tried int first. Didn’t work. The Hilbert implementation relies on floating point numbers. While it may be possible to convert it to fixed point numbers I didn’t really feel like doing it.

marciot replied:
I believe the only fractional number it uses is 0.5, so one could use a single bit for the decimals. But again, this would require some work.

X-Ryl669 replied:
No worry, this is a very good work already. The concerns I have is that it's a recursive function with the Hilbert level, so it'll overflow the stack if the subdivision is too high, and many (if not most) platform don't support floating point natively (and Hilbert's curves are typically implemented with fixed point usually, since, as you said, it's a clever division by a 1/4).

I've once implemented a space partitioning system based on Hilbert, that was using this trick to avoid using floating point value and fixed point value:
```c
/*
  This method was used to test the method above.
  It is twice as fast as AJ Yzelman's thesis version, because of shifts instead of divisions
*/
inline int computeHilbertsCurveValueFast(int x, int y, int logLevel) {
  static int HilbertRotationTable[4]  = { 3, 0, 0, 1};
  static int HilbertSenseTable[4]     = {-1, 1, 1, -1};
  static int HilbertQuadTable[4][2][2]= { {{0,1},{3,2}},
                      {{1,2},{0,3}},
                      {{2,3},{1,0}},
                      {{3,0},{2,1}} };

  int rotation = 0;
  int sense    = 1;
  int quad     = 0;

  int num      = 0;
  int k        = 0;
  int xbit     = 0;
  int ybit     = 0;


  for (k = logLevel - 1; k > 0; --k) {
    xbit = x>>k;
    ybit = y>>k;
    x -= xbit<<k;
    y -= ybit<<k;
    quad = HilbertQuadTable[rotation][xbit][ybit];
    num += ((sense == -1) ? (3-quad) : quad) << (k*2);
    rotation += HilbertRotationTable[quad];
    if (rotation >= 4) rotation -= 4;
    sense *= HilbertSenseTable[quad];
  }
  return num;
}
```

marciot replied:
Wow. This implementation is even more difficult to understand than the one I used. I wouldn’t mind it, except that it goes the wrong way. It generates an 1D coordinate from a 2D coordinate. I need the opposite.

I understand your concern about blowing the stack. I wrote this for a Due based board, it may not be appropriate for 8-but micros. I can probably convert to fixed point numbers somewhat easily. That would partially mitigate the issue. I’ll see if I can do that.

X-Ryl669 replied:
(Don't try to understand the output here, as I was generating an infinite length index using 2 bits per Hilbert Level). The trick I was try to show here is to shift left, perform the Hilbert curve 90° rotation and shift right just after to find the next coordinate, so that the coordinate can still be a integer.

marciot replied:
Ah. Okay. I get it. I’ve used fixed point math a lot in my life. I may write a to_fix and from_fix constexpr function for legibility, though.

marciot committed:
Use fixed point math when generating Hilbert curve

- Reduce stack memory consumpution four-fold, while supporting meshes up to 64x64

thinkyhead committed:
some formatting

thinkyhead committed:
ternify

thinkyhead committed:
etc.

thinkyhead committed:
Config option UBL_HILBERT_CURVE

thinkyhead committed:
add a test

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/21387

marciot commented:
@X-Ryl669: I’ve been able to convert this to fixed point math. Since Scott merged this, I’ll make a separate PR for that enhancement.

Integrating with G26 is proving to be way harder and I don’t know if and when I’ll be able to do it.
